### ISSUE TITLE
Double Image/Stuttering when reprojection is allowed

### ISSUE THREAD

**Issue by mapotter99** [LibreVR/Revive#251]
I think I have solved the Double Image problem, even with the latest nVidia drivers!!!

Open the SteamVR Settings, go to Performance, and UNcheck "Allow Reprojection".

After that, ALL of my double image problems went away!

Of course, you will probably want to re-enable it when playing native SteamVR games. It fills in missing FPS if the frame rate gets too low.

**Comment by kujohn** [comment]
This works for me! I wonder if it's Rift timewarp + reprojection is causing the issue.

**Comment by LibreVR** [comment]
Let me know if you still have problems with reprojection in [the latest release](https://github.com/LibreVR/Revive/releases/tag/0.9).

**Comment by mapotter99** [comment]
It is certainly improved!

Showdown Demo still has a tiny amount of double image... WAY better than before, but it looks somehow softer... probably due to the the tiny amount of double image.

Through the Ages National Parks has some significant frame rate issues... even more so with reprojection turned off (Perhaps an issue specifically with this particular experience).

Oculus Dreamdeck is completely smooth... finally!

Overall, I'd say it is 90-95% better, but there is still a little room for improvement.

Well done!

**Comment by mapotter99** [comment]
Just tried Damaged Core.  There is a slight double image with reprojection on.  It goes away with reprojection off.  Nowhere near as apparent before 0.90.

**Comment by LibreVR** [comment]
Thanks for testing, it's good to know that the tracking latency issue is related.

**Comment by JediDiggler** [comment]
I had the double image effect happen when playing Warthunder on Steam, which is weird as it's not an Oculus exclusive game. Could this be from just having Revive installed? Does seem to kick in whenever dropping below 90 fps and reprojection kicks in as mentioned. I think I prefer it to stuttering though, less headachy!

**Comment by LibreVR** [comment]
@JediDiggler This is a general symptom of the re-projection in SteamVR, it is **not** a Revive-specific issue. I can't even begin to explain what's wrong with your reasoning, but I'll try:

> I had the double image effect happen when playing Warthunder on Steam, which is weird as it's not an Oculus exclusive game.

Revive does not do anything in the background besides handling the dashboard. It will only affect games it's injected into by either launching Oculus Store titles through the dashboard or using the injector manually.

Please don't make such wild claims or other people may completely misunderstand what Revive is doing.

**Comment by JediDiggler** [comment]
Sorry, I didn't mean any offense, I thought the double image issue was from Revive/Oculus specific games, I'd never seen it in any Vive games until this. Apologies if I sounded accusational, I promise it wasn't intended!

**Comment by iamwes** [comment]
Mine has the double image / stutter if reprojection is on or off, even tried the newest driver from September and its the same. Had to roll back again to July's driver and its perfect... annoying though as some non vr games run better with the new driver.

**Comment by LibreVR** [comment]
Can you try playing with the following option enabled: `Steam VR Settings -> Developer -> Always On Reprojection`? It should help a lot for Revive game as it turns off Late Start which is a major bottleneck for Revive right now.

You should disable the option again for non-Revive games though.

**Comment by mapotter99** [comment]
@LibreVR, I can confirm that "Always On Reprojection" resolves the issues with revive 0.90. 

I just tried Revive 0.96 and am having several problems, including severe double issue, and Showdown not working in VR at all.  I'll test this a bit more and give more specific feedback about 0.96.

**Comment by mapotter99** [comment]
@LibreVR , I've installed 0.97.

Showdown is now working.  Thanks for fixing that.

I apologize for stating that 0.96 was "buggy" in another thread without further information.  I've been away with work since Friday with the exception of a few hours on Monday, and I wanted to try it out a bit more before going into details.

The main problem I'm having is the full return of the tracking latency/double image when turning my head or moving.  Revive 0.90 had gone a long way to fix that, but it is now as bad as ever with 0.97.  Disabling reprojection still resolves the issue, but I didn't see where "Always On Reprojection" resolved anything.

Also, Oculus video tracking latency is bad, but not in the same way.  When moving, the video is jumpy, rather than smeary like it is in most rendered games.

Is the tracking latency a result of the Oculus runtime, Revive, video drivers, or some combination of the above?  For what its worth, I have a GTX1070 with an i7, NVMe m.2 SSD, and 32 GB of Ram... I don't think system performance should be too much of a factor here.

Thanks for your hard work!

**Comment by LibreVR** [comment]
It's a combination of multiple factors, here's an overview of the issues we're dealing with though I'm still researching these so there may be some mistakes:

1. Prediction issues: The Oculus API allows the game to predict the pose an arbitrary amount of frames ahead. While OpenVR only allows you to predict 2 frames ahead if you want to keep in line with the compositor.
2. Latency issues: Because OpenVR doesn't use timewarp it wants the CPU to submit a frame as late as possible, also known as "late start". Oculus games don't expect that and may be doing some extra processing on the video thread that could've been done before we started waiting on the "late start".
3. Performance issues: Oculus games seem to expect that it can queue ahead a couple of frames on the CPU and then have timewarp correct the orientation. OpenVR doesn't allow that as timewarp can not handle big changes in translation.

These are difficult issues to solve and minor changes in the pipeline will affect these factors. For example the difference between 0.9.0 and 0.9.5 is that the former version made another copy of the frame before sending it to OpenVR. This could've affected either issue 1) or 2) though it's difficult to know for sure.

In any case I'm going to familiarize myself with all the documentation available around these problems and hopefully I'll have a fix before Revive 1.0.

**Comment by lghtanddrk** [comment]
If I just updated to 0.96 just to have the latest version would you suggest I skip this latest update?

**Comment by LibreVR** [comment]
0.9.6 is newer than 0.9.5, so it already includes those changes. From the games I've tested 0.9.5 is more compatible, though in some specific cases 0.9.0 may preform better.

Overall I recommend you use the latest version until we sort out the performance problems properly.

**Comment by Rabbitbeam** [comment]
I download the newest version and do a test.
I have not try revive with other game but only DragonFront.
It will appear double image if I check the reprojection.
Is it only a issue for DragonFront or every game should uncheck reprojection while playing by revive?

**Comment by mapotter99** [comment]
@LibreVR, I haven't had a chance to install the new SteamVR Beta with Asynchronus Reprojection yet, and it will be a couple days before I can test it. 

Should we expect Asynchronous Reprojection to help with these latency issues or make them worse?  Are there any compatibility issues we should be concerned about?

My apologies if you haven't had a chance to dig in to SteamVR's implementation of Asynchronous Reprojection yet.

**Comment by LibreVR** [comment]
It should definitely help, even without Async Reprojection there have been a bunch of changes in the SteamVR beta that improve motion-to-photon latency with Revive.

**Comment by LibreVR** [comment]
Also the new SteamVR beta fixes a bug with `PostPresentHandoff()` now I can get a *huge* performance boost in some games. In Dirt Rally we're talking about twice the frames per second, meaning you can finally play it on high with 4xMSAA at 90fps on a GTX1080.

**Comment by mapotter99** [comment]
@LibreVR , I've installed the latest SteamVR beta, the latest nVidia drivers, and the Revive 1.0 preview.

Oculus Dreamdeck is working beautifully!  No lag or latency issues at all!

Showdown is the best it has looked with reprojection, but it still has a very subtle latency lag...WAY better than before!  Interestingly, turning OFF all reprojection makes the latency far worse now!

Finally, Through the Ages: President Obama Celebrates the National Parks is still terribly jumpy.  There is no improvement, and it is arguably worse.  This is a 360 3D video... not sure if this issue is with the way Revive renders Oculus videos, or if it is program specific.

That's all I've had a chance to try for now.  If I notice anything else, I'll report back!

Thanks again!

**Comment by LibreVR** [comment]
I think this problem is now fixed in the latest version of SteamVR. There are still some jittering bugs in some OpenGL applications such as Through the Ages and Oculus Video, but that's a separate issue.
